[OP-52] Add boxcar filter.

mjwoudstra opened the pull request.

peendebak commented on src/qtt/algorithms/generic.py:
```diff
@@ -581,10 +585,47 @@ def weightedCentroid(im, contours, contourIdx, fig=None):
+
+
+    Args:
+        signal (sequence): a sequence containing the signal to be filtered.
```
```suggestion
        signal (array): An array containing the signal to be filtered.
```
The term sequence already as other meanings. Also I would not call a 2D array a 'sequence'

mjwoudstra replied:
Fine with me.

peendebak commented on src/qtt/algorithms/generic.py:
```diff
@@ -581,10 +585,47 @@ def weightedCentroid(im, contours, contourIdx, fig=None):
+        raise RuntimeError('Number of dimensions of kernel (%d) not equal to number of dimension of input signal (%d)' %
+                           (len(kernel_size), len(signal.shape)))
+
+    for siz in kernel_size:
```
Replace the next two lines with:
```
if np.any(kernel_size<=0):
   raise ...
```

mjwoudstra replied:
Ah, those fancy numpy functions

mjwoudstra replied:
Just need to convert the kernel_size to a np.array first

peendebak commented on src/qtt/algorithms/generic.py:
```diff
@@ -581,10 +585,47 @@ def weightedCentroid(im, contours, contourIdx, fig=None):
 # %%
 
 
 if __name__ == '__main__':
```
Remove these two lines, they do not do anything

mjwoudstra replied:
Agreed!

peendebak commented on src/tests/algorithms/test_generic.py:
```diff
@@ -40,3 +40,169 @@ def test_rescale_image(self):
+                msg = '\nsignal: %s\ngives wrong filtered output:' % (signal,)
+                msg += str(e).replace('x:', '  actual:').replace('y:', 'expected:')
+
+            if msg:
```
You can remove the `if msg:` by moving the statement into the `except` part.

mjwoudstra replied:
That is true from the program logic point of view and that is what it was initially. However, it turns out that if you do that, you get a warning that an exception was raised (by self.fail()) inside an exception handler. It does not seem to like that.

peendebak commented on src/tests/algorithms/test_generic.py:
```diff
@@ -40,3 +40,169 @@ def test_rescale_image(self):
+        kernel_2D = TestBoxcarFilter._kernel_size_2D
+        for signal, out_exp in TestBoxcarFilter._known_inputs_outputs_2D:
+            int_signal = np.array(signal, dtype=int)
+            try:
```
Replace the `try-except` block with
```
with self.assertRaises(AssertionError):
   np.testing.assert_allclose(int_signal, signal)
```
(similar for some other methods)

mjwoudstra replied:
The goal of the code is to only process those arrays whose float values are actually integers. The suggested replacement does not replicate that behavior. I have replaced the `np.testing.assert_allclose() + try/except` 
with 
`if np.allclose():`

peendebak commented:
@mjwoudstra Fixed some more issues with the new qcodes version tag and rebased to dev.

peendebak commented:
@CumulonimbusCalvus Fixes issues with the new mypy and the new version tag of qcodes. Can you review that part?

mjwoudstra commented:
> 
> 
> @mjwoudstra Fixed some more issues with the new qcodes version tag and rebased to dev.

@peendebak  That rebase also wiped out my TestBoxcarFilter test class. I put it back in commit 3e2b038.
In commit 2d6831c   I implemented a workaround for the packaging dependency suggested on qcodes slack by jenshnielsen this morning:
https://qcodes.slack.com/archives/C18KR2HD3/p1563953276010500

CumulonimbusCalvus commented:
@mjwoudstra @peendebak 

I moved the solution for #604 to PR #611, since this problems has nothing to do with adding a boxcar filter. Once done I will S&M, rebase and revert the updates related to #604.

CumulonimbusCalvus commented:
@mjwoudstra Should I rebase this branch or can you do it?

mjwoudstra commented:
@CumulonimbusCalvus I will rebase it to dev.

peendebak commented:
@mjwoudstra Can you add a single line to https://github.com/QuTech-Delft/qtt/blob/dev/CHANGELOG.md in this PR? 

I think we can merge then.
